toe/src/urbit/app/toe.hoon

```diff
@@ -0,0 +1,921 @@
+    |=  [per=player =board-game]
+    ^-  outcome
+    ?:  (wins per board-game)
+      `%wins
+    ?.  (ties board-game)
+      ~
+    `%tie
+  ::
+  ++  wins
+    |=  [per=player =board-game]
+    ^-  ?
+    %+  lien  winning-rows
+    |=  a=(list spot)
+    ^-  ?
+    %+  levy  a
+    |=  b=spot
+    =/  p=(unit player)  (~(get by board-game) b)
+    ?~(p %.n =(stone.per stone.u.p))
+  ::
+  ++  ties
+    |=(=board-game =(~(wyt in board-game) 9))
+  ::
+  ++  winning-rows
+    ^-  (list (list spot))
+    :~  ~[[%1 %1] [%2 %1] [%3 %1]]
+        ~[[%1 %2] [%2 %2] [%3 %2]]
+        ~[[%1 %3] [%2 %3] [%3 %3]]
+        ~[[%1 %1] [%1 %2] [%1 %3]]
+        ~[[%2 %1] [%2 %2] [%2 %3]]
+        ~[[%3 %1] [%3 %2] [%3 %3]]
+        ~[[%1 %3] [%2 %2] [%3 %1]]
+        ~[[%1 %1] [%2 %2] [%3 %3]]
+    ==
+  --
+::  +connect: handles subscriptions and updates
+::
+++  connect
+  |=  pax=path
+  ^-  (quip card _state)
+  ~&  who-connect+pax
+  |^  ?+    pax  ~|([%peer-toe-strange pax] !!)
+          [%sole *]           sole
+          [%toetile ~]        tile
+          [%room ^]           (room i.t.pax)
+          [%http-response *]  [~ state]
+      ==
+  ::  +tile: connects to the frontend
+  ::
+  ++  tile
+    ^-  (quip card _state)
+    ~&  tiled+src.bol
+    :_  state
+    [%give %fact ~ %json !>(*json)]~
+  ::  +sole: innitializes the console's state
+  ::
+  ++  sole
+    ^-  (quip card _state)
+    ~&  sole+[me src.bol]
+    ?.  (team:title me src.bol)  ~|([%strange-sole src.bol] !!)
+    :_  state(state.cli *sole-share)
+    =,  co:view
+    [(send mor+~[clear welcome wopr grid shall-we (prompt choose)])]~
+  ::  +room: receives a subscription to connecto to a room
+  ::
+  ++  room
+    |=  id=@t
+    ^-  (quip card _state)
+    =,  room-core
+    ?~  rooms  [~ state]
+    =/  ze=@p  (slav %p id)
+    ~&  [ze ship:current]
+    ?.  =(ze ship:current)  [~ state]
+    ?:  =(%wait game)
+      (join ze)
+    ?.  =(%play game)  [~ state]
+    (start ze)
+  --
+::  +view
+::  TODO: please make me into a %view app!
+::
+++  view
+  |%
+  ::  TODO: move other view updates here
+  ::
+  ++  updates
+    |%
+    ++  restart
+      ^-  (list card)
+      :~  (send:co mor+all-effects:co)
+          (send:fe [%status s+'null']~)
+      ==
+    ::
+    ++  receive
+      |=  guest=@p
+      =/  room-name=tape  (cite:title guest)
+      ?^  rooms
+        (notify guest)
+      :~  (send:fe ~[[%message s+(crip room-name)] [%status s+'confirm']])
+          =,  co
+          (send (prompt "{confirm}{room-name}? (Y/N) | "))
+      ==
+    ::
+    ++  confirm
+      =/  next=@p  guest:next:room-core
+      :~  (confirm:updates:fe next)
+          (confirm:updates:co next)
+      ==
+    ::
+    ++  playing
+      |=  move=toe-turno
+      ^-  (list card)
+      =/  [ze=@p who=@p]   =,  current:room-core
+        [ship who]
+      :~  ::  If the move originated in the frontend, this is redundant
+          ::  so this move is ignored on the frontend (via %next)
```

Comment:
Not sure I quite get this yet, but seems like a smell? Might be fine though.

Reply:
Yeah, this is from the previous version, I want to make sure that it still applies... be back in a sec!

Reply:
So, if we input a move via the console, we update the board state and send a card to our opponent and to our browser. 

If we put a move from the frontend, we poke our app with a json payload, put it on the board, send the card to our opponent and to our fronte... ok, this is redundant. but! with the next flag that we send for every move, we know that it's not our turn, so no harm is done in the frontend and the state is the same.

Reply:
I see. Sending updates to all frontends always is perfectly sane. Carry on!